src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeNameParserHelpers.cs

```diff
@@ -102,29 +102,54 @@ static int GetUnescapedOffset(ReadOnlySpan<char> input, int startOffset)
             static bool NeedsEscaping(char c) => c is '[' or ']' or '&' or '*' or ',' or '+' or EscapeCharacter;
         }
 
-        internal static ReadOnlySpan<char> GetName(ReadOnlySpan<char> fullName)
+        internal static ReadOnlySpan<char> GetNamespace(ReadOnlySpan<char> fullName)
         {
-            // The two-value form of MemoryExtensions.LastIndexOfAny does not suffer
-            // from the behavior mentioned in the comment at the top of GetFullTypeNameLength.
-            // It always takes O(m * i) worst-case time and is safe to use here.
+            // Detect if a nested type name has a namespace (e.g. "a.b+c") and throw if it does.
+            while (LastIndexOfUnescaped(fullName, '+') is int nestingBoundaryPos && nestingBoundaryPos > 0)
+            {
+                if (LastIndexOfUnescaped(fullName.Slice(nestingBoundaryPos + 1), '.') > 0)
+                {
+                    ThrowInvalidOperation_NestedTypeNamespace();
+                }
+                fullName = fullName.Slice(0, nestingBoundaryPos);
+            }
+            int offset = LastIndexOfUnescaped(fullName, '.');
+            return offset < 0 ? [] : fullName.Slice(0, offset);
+        }
 
-            int offset = fullName.LastIndexOfAny('.', '+');
+        internal static ReadOnlySpan<char> GetName(ReadOnlySpan<char> fullName)
+        {
+            // If the type is nested, return the whole name after the plus sign.
```

Comment:
This will be do more work than before for non-nested types - it will need to go through the full name instead of stopping at the first dot. Would it be better if the caller passed in the information about whether the type is nested or not to avoid that?

Reply:
Added an `IsNested` parameter and updated uses.